btree/btree.c

```diff
@@ -1885,12 +1892,44 @@ static int64_t mem_tree_get(struct node_op *op, struct segaddr *addr, int nxt);
  */
 static int64_t tree_get(struct node_op *op, struct segaddr *addr, int nxt)
 {
-	int        nxt_state;
+	struct td              *tree = NULL;
+	struct nd              *node = NULL;
 
-	nxt_state = segops->so_tree_get(op, addr, nxt);
+	if (addr != NULL) {
+		nxt  = node_get(op, NULL, addr, nxt);
+		if (op->no_op.o_sm.sm_rc < 0)
+			return op->no_op.o_sm.sm_rc;
+		node = op->no_node;
+		tree = node->n_tree;
 
-	return nxt_state;
+		if (tree == NULL) {
+			tree = m0_alloc(sizeof *tree);
+			m0_rwlock_init(&tree->t_lock);
+			m0_rwlock_write_lock(&tree->t_lock);
+
+			tree->t_ref = 1;
+			tree->t_starttime = 0;
+			tree->t_root = node;
+			tree->t_height = node_level(node) + 1;
+			node_lock(node);
+			node->n_tree = tree;
+			node_unlock(node);
+
+		} else {
+			m0_rwlock_write_lock(&tree->t_lock);
+			tree->t_ref++;
```

Comment:
It may happen by the time you reach this point tree descriptor may become freed by other thread executing tree_put operation. So, in this case you are trying to take on lock which doesn't exists.

Better to use global lock instead of tree_lock and this global locks need to be taken at the beginning of tree_get and tree_put operation and released at the end of these operation, In this way you can avoid above scenario.

Reply:
Ok Yes.
I will add that lock!